### PR TITLE
Enter Access Code

### PR DESCRIPTION
**Pull request by chrisrcoles**
## Description

This PR introduces functionality that requires service members to enter a valid access code before they start processing their move. An access code is comprised of either `HHG` or `PPM`, along with an alpha-numeric character set of 6, e.g., `HHG-D34LKJ`. Once service members enter the access code, it will be validated and then claimed (associated with the service member) and they will then move on to the first step - setting up their service member profile. 

## Reviewer Notes

Please see inline comments for thoughts.

This work is dependent upon two outstanding PRs:
1. [Claim Access Code](https://github.com/transcom/mymove/pull/2129) 
2. [Redirect User to Access Code Page](https://github.com/transcom/mymove/pull/2169)

Please note, both of these PRs should be resolved prior to merging this PR.

**It is important to note that once this code is merged in, all new users will be required to enter a valid access code. This should not be merged in until Product - specifically, Linda Setchell and Erin Stanfill have given the green light.** 

## Setup

1. `make server_run`
2. `make client_run`
3. Generate an access code `go run cmd/generate_access_codes/main.go -hhg 1`
4. Query the DB for the created access code. 
5. Sign in to app, either through `Local Sign In` or `Sign In`. Whichever manner you use to sign in, you must have a completely new user. For `Local Sign In` option, click `Create MilMoveUser`, for `Sign In` option, create an entirely new Login.gov account.
6. Log In and you will be redirected to access code page.
7. Enter access code.

This work also includes handling the common different error conditions for failure - `500`, `invalid format`, and `invalid`.

## Code Review Verification Steps

* [ ] Design Review
* [x] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [x] User facing changes have been reviewed by design.
* [x] Request review from a member of a different team.
* [x] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/163338921)

### CODE REVIEW

File: pkg/handlers/internalapi/users_test.go [hunk 1]
```diff
@@ -43,13 +46,28 @@ func (suite *HandlerSuite) TestServiceMemberLoggedInUserHandler() {
...
+		MoveType: &selectedMoveType,
+	}
+	suite.MustSave(&accessCode)
+
+	accessCodeFetcher := &mocks.AccessCodeFetcher{}
+	accessCodeFetcher.On("FetchAccessCode",
+		mock.AnythingOfType("UUID"),
+	).Return(&accessCode, nil)
+
+	handler := ShowLoggedInUserHandler{handlers.NewHandlerContext(suite.DB(), suite.TestLogger()), accessCodeFetcher}
```
Comment (kahlouie) [inline pkg/handlers/internalapi/users_test.go]: Instead of testing service member log in only with an access code, could we also test login when a user doesn't require an access code, so we know if we break existing users? (PS already tested that manually, but want to make sure we don't accidentally break it later on. 😃)

File: cypress/integration/mymove/serviceMemberProfile.js [hunk 2]
```diff
@@ -1,16 +1,26 @@
 /* global cy */
+
```
Comment (kahlouie) [inline cypress/integration/mymove/serviceMemberProfile.js]: Can we also test a service member who doesn't require an access code?

File: pkg/auth/authentication/devlocal.go [hunk 3]
```diff
@@ -62,7 +62,7 @@ func (h UserListHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			http.StatusInternalServerError)
 		return
 	}
-	// Truncate the list if larger than 25
+	// // Truncate the list if larger than 25
```
[the remark this answers is not in the document]
Comment (ralren) [reply pkg/auth/authentication/devlocal.go]: Whoops. Taking that off.

File: pkg/handlers/internalapi/users_test.go (hunk 1 above)
Comment (ralren) [reply pkg/handlers/internalapi/users_test.go]: The next test should handle that but it was broken anyway. I added a new test and fixed the next one.

File: cypress/integration/mymove/serviceMemberProfile.js (hunk 2 above)
Comment (ralren) [reply cypress/integration/mymove/serviceMemberProfile.js]: I made a go at this but it requires some extra steps that I don't think is worth the trouble. I'll pass on this.

File: pkg/handlers/internalapi/service_members.go [hunk 4]
```diff
@@ -146,7 +149,7 @@ func (h CreateServiceMemberHandler) Handle(params servicememberop.CreateServiceM
 		session.LastName = *(newServiceMember.LastName)
 	}
 	// And return
-	serviceMemberPayload := payloadForServiceMemberModel(h.FileStorer(), newServiceMember)
+	serviceMemberPayload := payloadForServiceMemberModel(h.FileStorer(), newServiceMember, "")
```
Comment (sarboc) [inline pkg/handlers/internalapi/service_members.go]: Is there anyway to be more explicit about what the empty string means in this context? We could set a default in the method, or create a constant somewhere in this file, or even just a local named variable would make it more clear what's going on here.

File: pkg/handlers/internalapi/service_members.go [hunk 5]
```diff
@@ -56,6 +56,8 @@ func payloadForServiceMemberModel(storer storage.FileStorer, serviceMember model
 		HasSocialSecurityNumber: handlers.FmtBool(serviceMember.SocialSecurityNumberID != nil),
 		IsProfileComplete:       handlers.FmtBool(serviceMember.IsProfileComplete()),
 		CurrentStation:          payloadForDutyStationModel(serviceMember.DutyStation),
+		RequiresAccessCode:      serviceMember.RequiresAccessCode,
+		AccessCode:              accessCode,
```
Comment (sarboc) [inline pkg/handlers/internalapi/service_members.go]: Is there a reason we need to return the access code with the payload? Does the front end ever use it? I would assume that the front end would send it with some kind of initial request, but we wouldn't need to receive it back for any reason - just a success/failure of the request depending on whether or not the code was valid.

File: pkg/services/accesscode/fetch_access_code.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+func NewAccessCodeFetcher(db *pop.Connection) services.AccessCodeFetcher {
+	return &fetchAccessCode{db}
+}
+
+// FetchAccessCode fetches an access code based upon the service member id.
+func (f fetchAccessCode) FetchAccessCode(serviceMemberID uuid.UUID) (*models.AccessCode, error) {
+	ac := models.AccessCode{}
+	err := f.DB.
+		Where("service_member_id = ?", serviceMemberID).
+		First(&ac)
```
Comment (sarboc) [inline pkg/services/accesscode/fetch_access_code.go]: Will we only ever have one code per service member? If that's an expectation, should we add some kind of uniqueness constraint on the access codes table?

File: src/scenes/Admin/index.jsx [hunk 7]
```diff
@@ -9,7 +9,7 @@ import AdminHeader from 'shared/Header/Admin';
 import { getCurrentUserInfo } from 'shared/Data/users';
 import { no_op } from 'shared/utils';
 import LogoutOnInactivity from 'shared/User/LogoutOnInactivity';
-import PrivateRoute from 'shared/User/PrivateRoute';
+import ValidatedPrivateRoute from 'shared/User/ValidatedPrivateRoute';
```
Comment (sarboc) [inline src/scenes/Admin/index.jsx]: Small-ish refactors like this that change a lot of files/lines of code are good candidates to be pulled into a separate PR. This code could get merged without changing any user-facing functionality before the rest of the story is complete, and it would make the content of the final PR a) shorter and b) more relevant to the story at hand. Don't be afraid to do more than one PR for a story.

File: src/shared/Entities/modules/accessCodes.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+const claimAccessCodeLabel = 'AccessCodes.updateAccessCodeForServiceMember';
+
+export function validateAccessCode(code, label = validateAccessCodeLabel) {
+  const swaggerTag = 'accesscode.validateAccessCode';
+  return swaggerRequest(getPublicClient, swaggerTag, { code }, { label });
+}
+
+export function claimAccessCode(accessCode, label = claimAccessCodeLabel) {
+  const swaggerTag = 'accesscode.claimAccessCode';
+  return swaggerRequest(getPublicClient, swaggerTag, { accessCodePayload: accessCode }, { label });
```
Comment (sarboc) [inline src/shared/Entities/modules/accessCodes.js]: If you removed the term `Payload` at the swagger level, you a) wouldn't need to manually construct the object here (`{ accessCode }` would suffice), and b) I think you would quiet a likely console warning reminding you that `Payload` is deprecated.

File: src/shared/Entities/modules/accessCodes.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { swaggerRequest } from 'shared/Swagger/request';
+import { getPublicClient } from 'shared/Swagger/api';
+
+const validateAccessCodeLabel = 'AccessCodes.getAccessCode';
+const claimAccessCodeLabel = 'AccessCodes.updateAccessCodeForServiceMember';
+
+export function validateAccessCode(code, label = validateAccessCodeLabel) {
```
Comment (sarboc) [inline src/shared/Entities/modules/accessCodes.js]: This is a thing of beauty. Excellent, clear, swagger request function.

File: src/shared/User/AccessCode.css [hunk 10]
```diff
@@ -0,0 +1,15 @@
+.usa-input,
```
Comment (sarboc) [inline src/shared/User/AccessCode.css]: We have a css modules pattern that you should use here. Also, I would advise against modifying the base classes for the web standards input, since it might have unintended consequences. It would be more fail-safe to add an additional class to the inputs, rather than over-ride the base input classes.

As a side note, due to the way css imports work, these overrides would apply to every input field on a page where this css was imported, even if it was only imported in one React component on the page. Using the modules pattern limits the scope/bleed of the css.

File: src/shared/User/AccessCode.jsx [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  validateAndClaimAccessCode = () => {
+    const { formValues, validateAccessCode, claimAccessCode } = this.props;
+    const code = formValues.claim_access_code;
+    this.validateAccessCodePattern(code);
+
+    return validateAccessCode(code)
+      .then(res => {
+        const { body: accessCode } = get(res, 'response');
+        if (!accessCode.code) {
+          throw new SubmissionError({
```
Comment (sarboc) [inline src/shared/User/AccessCode.jsx]: Throwing errors after performing some swagger request options feels like we're creating new patterns, and I'm not sure this is the best route to take. If we want to perform actions, and store information in the reducer (using swagger request), then my hunch would be that we should continue/not continue in the flow based on information found in the reducer (either in the access code entity or in the request log), not based on throwing errors.

It might be worth looking into how we stop flow progression in other places in the service member app, depending on what the server response is.

File: src/shared/User/ValidatedPrivateRoute.jsx [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import SignIn from './SignIn';
+import AccessCode from './AccessCode';
+
+// this was adapted from https://github.com/ReactTraining/react-router/blob/master/packages/react-router-redux/examples/AuthExample.js
+// note that it does not work if the route is not inside a Switch
+class ValidatedPrivateRouteContainer extends React.Component {
+  render() {
+    const { isLoggedIn, requiresAccessCode, accessCode, path, ...props } = this.props;
+    if (!isLoggedIn) return <Route path={path} component={SignIn} />;
+    if (isLoggedIn && requiresAccessCode && !accessCode) return <Route path={path} component={AccessCode} />;
```
Comment (sarboc) [inline src/shared/User/ValidatedPrivateRoute.jsx]: Ah. I see now why we're returning the access code with the service member payload. Is there some other way we can logically determine whether a service member is logging in for the first time and needs to enter an access code to continue? 

Here is why I'm asking: The access code is important at sign up, but after that, I don't think we really care about it. All we care about is that the user has previously validated they have permission to create a move in the new system. So, continuing to check for the code every single time a service member logs in, in perpetuity, seems like a lot of checking for something we only care about at account creation.

There might be other ways to determine whether or not the service member is logging in for the first time, and use that logic to create a gate here. For example, if the service member requires an access code, and no move exists, then we probably need to run the access code check. If the service member requires an access code, but already has a move, it is likely safe to infer that we have already checked the access code for them.

File: swagger/internal.yaml [hunk 13]
```diff
@@ -2498,6 +2507,15 @@ definitions:
     type: string
     x-nullable: true
     title: Vehicle nickname (ex. "My car")
+  ClaimAccessCodePayload:
```
Comment (sarboc) [inline swagger/internal.yaml]: As your JS console will likely tell you... `'Payload' is deprecated`

File: src/shared/User/AccessCode.css (hunk 10 above)
Comment (sarboc) [reply src/shared/User/AccessCode.css]: More info on the css stuff here: https://github.com/transcom/mymove/blob/master/docs/frontend.md#styling-standards

File: migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz [hunk 14]
```diff
@@ -0,0 +1 @@
+add_column("service_members", "requires_access_code", "bool", {"default": false})
```
[the remark this answers is not in the document]
Comment (chrisrcoles) [reply migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz]: Hey @sarboc yes the default is proper here. Check out the design doc [here](https://docs.google.com/document/d/1D5f52XMC0PMdBaiv85I5W9nMxOTv0EU1mFsBMDqSXo0/edit) for more explanation.

File: pkg/handlers/internalapi/service_members.go (hunk 4 above)
Comment (chrisrcoles) [reply pkg/handlers/internalapi/service_members.go]: agreed - cc @ralren

File: pkg/handlers/internalapi/service_members.go (hunk 5 above)
Comment (chrisrcoles) [reply pkg/handlers/internalapi/service_members.go]: yes @sarboc , when a service member chooses their move type, they can only choose a move type that matches the approved access code so access codes will need to be stored in redux state.

File: pkg/services/accesscode/fetch_access_code.go (hunk 6 above)
Comment (chrisrcoles) [reply pkg/services/accesscode/fetch_access_code.go]: agreed - cc @ralren

File: src/scenes/Admin/index.jsx (hunk 7 above)
Comment (chrisrcoles) [reply src/scenes/Admin/index.jsx]: Agreed - there are about 4 PRs that have led up to this one. All have been either closed or are still open.

Comment (chrisrcoles) [reply src/scenes/Admin/index.jsx]: This [PR](https://github.com/transcom/mymove/pull/2169) captures the changes for `ValidatedPrivateRoute`

File: src/shared/Entities/modules/accessCodes.js (hunk 9 above)
Comment (chrisrcoles) [reply src/shared/Entities/modules/accessCodes.js]: :+1:

File: src/shared/User/AccessCode.css (hunk 10 above)
Comment (chrisrcoles) [reply src/shared/User/AccessCode.css]: agreed - cc @ralren :+1:

File: src/shared/User/ValidatedPrivateRoute.jsx (hunk 12 above)
Comment (ralren) [reply src/shared/User/ValidatedPrivateRoute.jsx]: We actually had a discussion earlier today to use the access code later down the road when the user selects a move type. Is this a good enough reason to keep it here (after being modified into the object and not just the code itself)?

File: migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz (hunk 14 above)
[the remark this answers is not in the document]
Comment (sarboc) [reply migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz]: @chrisrcoles The design doc says that the default value for this column should be `true`.

File: src/scenes/Admin/index.jsx (hunk 7 above)
Comment (sarboc) [reply src/scenes/Admin/index.jsx]: Ahhh. Gotcha. Thanks for the context.

File: migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz (hunk 14 above)
[the remark this answers is not in the document]
Comment (ralren) [reply migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz]: Despite what the design doc says, false is the proper default. When we migrate, we want to make sure all past users have this column set to false.

File: src/shared/User/ValidatedPrivateRoute.jsx (hunk 12 above)
Comment (sarboc) [reply src/shared/User/ValidatedPrivateRoute.jsx]: I'm still not sure I would use that in the access check, or return it as part of the service member payload. But let me think about that new wrinkle for a bit.

File: migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz (hunk 14 above)
[the remark this answers is not in the document]
Comment (sarboc) [reply migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz]: @ralren Handling previous records and how we handle new records might need to be tackled separately. If existing users are the reason for using `false` as the default, but we actually want the default to be `true`, then I would argue we should add the column, populate existing users, and then change the default to be the value we want moving forward.

If `false` is really the default we want for new users, though, then this is fine. It seemed odd to me that we were using a default of false, but expecting all new users to have a value of true.

File: src/shared/User/ValidatedPrivateRoute.jsx (hunk 12 above)
Comment (chrisrcoles) [reply src/shared/User/ValidatedPrivateRoute.jsx]: Yes - check the [design doc](https://docs.google.com/document/d/1D5f52XMC0PMdBaiv85I5W9nMxOTv0EU1mFsBMDqSXo0/edit).

File: swagger/internal.yaml (hunk 13 above)
Comment (chrisrcoles) [reply swagger/internal.yaml]: cc - @ralren

File: migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz (hunk 14 above)
[the remark this answers is not in the document]
Comment (ralren) [reply migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz]: We'll be handling what value new users will have through a feature flag in this PR: https://github.com/transcom/mymove/pull/2230

File: pkg/services/accesscode/fetch_access_code.go (hunk 6 above)
Comment (ralren) [reply pkg/services/accesscode/fetch_access_code.go]: Done

File: src/shared/User/ValidatedPrivateRoute.jsx (hunk 12 above)
Comment (ralren) [reply src/shared/User/ValidatedPrivateRoute.jsx]: I'm tabling this for now. The goal right now is to get this into staging asap. There is now a change to get the access code information through a new endpoint. I'll be implementing this in the second pass and I'll be modifying the logic so that access code is no longer needed here.

File: pkg/handlers/internalapi/service_members.go (hunk 5 above)
Comment (ralren) [reply pkg/handlers/internalapi/service_members.go]: Tabling this. Left a comment about this in the discussion within `ValidatedPrivateRoute.jsx`

File: swagger/internal.yaml (hunk 13 above)
Comment (ralren) [reply swagger/internal.yaml]: Modified this.

File: src/shared/User/AccessCode.jsx (hunk 11 above)
Comment (ralren) [reply src/shared/User/AccessCode.jsx]: If it's okay with you, I would like to handle this in a second pass as well.

File: src/shared/User/ValidatedPrivateRoute.jsx (hunk 12 above)
Comment (sarboc) [reply src/shared/User/ValidatedPrivateRoute.jsx]: @ralren That sounds like an excellent plan. Let me know if you need any help as you make those changes.

File: src/shared/User/AccessCode.css (hunk 10 above)
Comment (ralren) [reply src/shared/User/AccessCode.css]: Fixed this.
